apps/src/p5lab/spritelab/commands/validationCommands.js

```diff
@@ -57,10 +64,583 @@ export const commands = {
     };
   },
 
+  // Gets an array of ids for any sprite that has an associated event triggered this frame.
+  getEventSpriteIds() {
+    // We want to store any ids that are included in events logged this frame.
+    // Touch events include two distinct sprite ids.
+    let idArray = [];
+
+    // Only check for values that are new this frame.
+    for (let i = this.previous.eventLogLength; i < this.eventLog.length; i++) {
+      if (
+        // Check for each event type that includes sprite ids (NOT time or key events).
+        this.eventLog[i].includes('whenClick: ') ||
+        this.eventLog[i].includes('whileClick: ') ||
+        this.eventLog[i].includes('whenTouch: ') ||
+        this.eventLog[i].includes('whileTouch: ') ||
+        this.eventLog[i].includes('spriteCreated: ')
+      ) {
+        // Use .concat because it's possible for multiple events to be logged in the same frame.
+        idArray = idArray.concat(
+          // Take whatever was in the current eventLog entry...
+          this.eventLog[i]
+            // ...remove the spaces and create an array...
+            .split(' ')
+            // ...convert each string in the array to a number...
+            .map(Number)
+            // ...remove NaN values. (ex. 'whenClick: 0' results in [Nan, 0] above)
+            .filter(function(value) {
+              return !Number.isNaN(value);
+            })
+        );
+      }
+    }
+    return idArray;
+  },
+
+  // Used in levels (typically first frame only) to create an ordered list of success criteria
+  addCriteria(predicate, feedback) {
+    if (typeof predicate === 'function' && typeof feedback === 'string') {
+      this.criteria.push(new criteria(predicate, feedback));
+    }
+  },
+
+  // Used in levels to override default validation timing.
+  setEarlyTime(frames) {
+    this.validationTimes.early = frames;
+  },
+  setWaitTime(frames) {
+    this.validationTimes.wait = frames;
+  },
+  setDelayTime(frames) {
+    this.validationTimes.delay = frames;
+  },
+
+  // Used in levels (typically every frame) to validate based on all criteria
+  validate() {
+    // Get the current (ie. previous frame) pass/fail state prior to validation
+    const state = commands.getPassState(this.criteria);
+
+    const barWidth =
+      this.currentFrame() * commands.calculateBarScale(this.validationTimes);
+    drawUtils.validationBar(this.p5, barWidth, state, {});
```

Comment:
I feel like we could separate out the job of doing validation from rendering some UI.  If we can make `validate` a pure function -- i.e. it takes some input and returns a value -- then elsewhere we could update the UI based on that information.

Reply:
On the other hand, maybe this function does need to modify some things, in which case maybe it could be called 'updateValidation' or something to imply that it takes external information and updates UI & other things?

Reply:
I went with the latter suggestion and added more comments to explain the design.